Crystal 0.23.1 can't be built with Crystal 0.24.x

## Problem

Currently, `./install.sh 0.23.1` for Crystal fails with a compilation error. I know it succeeded in the past, but now it doesn't.

## How to reproduce

```
wandbox-builder/build$ sudo ./docker-run.sh crystal
/var/work/crystal# ./install.sh 0.23.1

(omit some lines here)

+ cd crystal-0.23.1/
+ make -j2 verbose=1
Using /usr/bin/llvm-config-3.8 [version=3.8.0]
g++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc `/usr/bin/llvm-config-3.8 --cxxflags`
cc -fPIC    -c -o src/ext/sigfault.o src/ext/sigfault.c
ar -rcs src/ext/libcrystal.a src/ext/sigfault.o
CRYSTAL_CONFIG_PATH=`pwd`/src ./bin/crystal build  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib
Error in src/compiler/crystal.cr:6: while requiring "./crystal/**"

require "./crystal/**"
^

in src/compiler/crystal/formatter.cr:10: while requiring "./tools/formatter"

require "./tools/formatter"
^

in src/compiler/crystal/tools/formatter.cr:1: while requiring "../syntax"

require "../syntax"
^

in src/compiler/crystal/syntax.cr:10: while requiring "./syntax/**"

require "./syntax/**"
^

Syntax error in src/compiler/crystal/syntax/parser.cr:3957: duplicate when :+ in case

      when :CHAR, :STRING, :DELIMITER_START, :STRING_ARRAY_START, :SYMBOL_ARRAY_START, :NUMBER, :IDENT, :SYMBOL, :INSTANCE_VAR, :CLASS_VAR, :CONST, :GLOBAL, :"$~", :"$?", :GLOBAL_MATCH_DATA_INDEX, :REGEX, :"(", :"!", :"[", :"[]", :"+", :"-", :"~", :"&", :"->", :"{{", :__LINE__, :__END_LINE__, :__FILE__, :__DIR__, :UNDERSCORE
                                                                                                                                                                                                                                      ^

Makefile:116: recipe for target '.build/crystal' failed
make: *** [.build/crystal] Error 1
```

## Workaround & Proposal

It seems that this error is caused by compiling Crystal by a newer version of Crystal compiler. Currently, `apt install crystal` installs Crystal 0.24.1, and the build fails with it. (It also fails with 0.24.0.) However, the build succeeds with manually-installed Crystal 0.22.0. So this problem might be caused by the upgrade of the Crystal compiler installed by `apt`.

Therefore, how about installing an old compiler from <https://github.com/crystal-lang/crystal/releases>? I don't know if it's OK for the Wandbox installation policy, but if this is OK I will be able to send a pull req.

## Environment

* Host: Ubuntu 17.10
* Docker 17.12.0-ce
* Wandbox-builder a216d20dad9eb3aea73e56286f4bf46c918133a1

--- a/build/crystal/install.sh
+++ b/build/crystal/install.sh
@@ -10,6 +10,21 @@ fi
 VERSION="$1"
 PREFIX="/opt/wandbox/crystal-$VERSION"
 
+cd /usr/local/
+
+# prepare for bootstrapping
+if compare_version "$VERSION" "<=" "0.23.1" ; then
+  # Crystal 0.23.1 can't be compiled by Crystal >= 0.24.0 (melpon/wandbox-builder#17)
+  archive="crystal-0.23.1-3-linux-x86_64.tar.gz"
+  wget_strict_sha256 \
+    "https://github.com/crystal-lang/crystal/releases/download/0.23.1/${archive}" \
+    "$BASE_DIR/resources/${archive}.sha256"
+  tar xf "./${archive}"
+  rm "./${archive}"
+  export PATH="/usr/local/crystal-0.23.1-3/bin/:$PATH"
+fi
+crystal --version
+
 cd ~/
 
 # get source

--- a/build/crystal/resources/crystal-0.23.1-3-linux-x86_64.tar.gz.sha256
+++ b/build/crystal/resources/crystal-0.23.1-3-linux-x86_64.tar.gz.sha256
@@ -1,0 +1,1 @@
+6a84cc866838ffa5250e28c3ce1a918a93f89c06393fe8cfd4068fcbbc66f3ab *crystal-0.23.1-3-linux-x86_64.tar.gz

